src/components/ErrorDecoder.tsx

```diff
@@ -0,0 +1,94 @@
+import {useMemo} from 'react';
+import {useRouter} from 'next/router';
+
+function replaceArgs(
+  msg: string,
+  argList: Array<string | undefined>,
+  replacer = '[missing argument]'
+): string {
+  let argIdx = 0;
+  return msg.replace(/%s/g, function () {
+    const arg = argList[argIdx++];
+    // arg can be an empty string: ?invariant=377&args[0]=&args[1]=count
+    return arg === undefined || arg === '' ? replacer : arg;
+  });
+}
+
+// When the message contains a URL (like https://fb.me/react-refs-must-have-owner),
+// make it a clickable link.
+function urlify(str: string): React.ReactNode {
+  const urlRegex = /(https:\/\/fb\.me\/[a-z\-]+)/g;
+
+  const segments = str.split(urlRegex);
+
+  return segments.map((message, i) => {
+    if (i % 2 === 1) {
+      return (
+        <a key={i} target="_blank" rel="noopener noreferrer" href={message}>
+          {message}
+        </a>
+      );
+    }
+    return message;
+  });
+}
+
+// `?invariant=123&args[]=foo&args[]=bar`
+// or `// ?invariant=123&args[0]=foo&args[1]=bar`
+function parseQueryString(
+  query: ReturnType<typeof useRouter>['query']
+): Array<string | undefined> {
+  const args: Array<string | undefined> = [];
+
+  Object.entries(query).forEach(([key, value]) => {
+    if (key.startsWith('args[')) {
+      args.push(Array.isArray(value) ? value[0] : value);
+    }
+  });
+
+  return args;
+}
+
+interface ErrorDecoderProps {
+  errorMessages: string | null;
+}
+
+export default function ErrorDecoder({errorMessages}: ErrorDecoderProps) {
+  const {isReady, query} = useRouter();
+
+  const msg = useMemo(() => {
+    if (!errorMessages?.includes('%s')) {
+      return errorMessages;
+    }
+
+    if (typeof window !== 'undefined' && isReady) {
+      return replaceArgs(
+        errorMessages,
+        parseQueryString(query),
+        '[missing argument]'
+      );
+    }
+
+    return replaceArgs(errorMessages, [], '[parsing argument]');
+  }, [errorMessages, isReady, query]);
```

Comment:
@rickhanlonii 

Turns out that Next.js SSG is fast enough. The total build time is at 95 seconds now.

Note that although the page itself has been SSG-ed at the build time, the `router.query` still requires client-side rendering and will only be populated after the initial hydration, thus `isReady` can not be avoided.

Reply:
When the query changes, won't that trigger an update? Why is the `isReady` needed?

Reply:
> When the query changes, won't that trigger an update? Why is the `isReady` needed?

Per Next.js documentation: https://nextjs.org/docs/pages/building-your-application/rendering/automatic-static-optimization#how-it-works

> To be able to distinguish if the query is fully updated and ready for use, you can leverage the `isReady` field on [`next/router`](https://nextjs.org/docs/pages/api-reference/functions/use-router#router-object).

This also prevents SSR hydration mismatch errors. On the server, Next.js SSR the page without the `query`, and Next.js will also use the empty query object during the initial hydration.

The `isReady: true` will only be set after the initial hydration has finished and the `query` object has been populated (the `isReady` here acts similarly with [React's two-pass rendering approach](https://react.dev/reference/react-dom/client/hydrateRoot#handling-different-client-and-server-content)).

The `useSearchParams` of the Next.js App Router uses the Suspense approach, which is useful when we migrate to the App Router.

Reply:
Ok that's a bummer. I also see that there are multiple updates that happen before `query` and `isReady` are set, delaying the time it takes to render the final message.

Instead, why don't we read the query params ourselves instead of from the router. That will allow us to set the message in an effect, which fires immediately after hydration, and that will set it at the first moment possible.

Finally, we can add a fade in animation to avoid the flash of unparsed args. Something like this:

```js
function parseQueryString(search: string): Array<string> {
  const rawQueryString = search.substring(1);
  if (!rawQueryString) {
    return [];
  }

  let args = [];

  const queries = rawQueryString.split('&');
  for (let i = 0; i < queries.length; i++) {
    const query = decodeURIComponent(queries[i]);
    if (query.indexOf('args[') === 0) {
      args.push(query.slice(query.indexOf(']=') + 2));
    }
  }

  return args;
}

export default function ErrorDecoder({errorMessages}: ErrorDecoderProps) {
  const [message, setMessage] = useState<string | null>(errorMessages);

  useEffect(() => {
    if (errorMessages == null) {
      return;
    }
    setMessage(
      replaceArgs(
        errorMessages,
        parseQueryString(window.location.search || ''),
        '[missing argument]'
      )
    );
  }, [errorMessages]);

  const displayMessage = useMemo(() => {
    if (!message) {
      return null;
    }
    return urlify(
      replaceArgs(message, parseQueryString(''), '[missing argument]')
    );
  }, [message]);

  if (!displayMessage) {
    return (
      <p>
        When you encounter an error, you{"'"}ll receive a link to this page for
        that specific error and we{"'"}ll show you the full error text.
      </p>
    );
  }

  return (
    <div>
      <p>
        <b>The full text of the error you just encountered is:</b>
      </p>
      <code className="block bg-red-100 text-red-600 py-4 px-6 mt-5 rounded-lg animate-fade-up">
        <b>{displayMessage}</b>
      </code>
    </div>
  );
}
```

```js
// add to tailwind.config.js
{
  theme: {
    extend: {
      animation: {
        'fade-up': 'fade-up 1s both',
        // ...
      } ,
      keyframes: {
        'fade-up': {
          '0%': {
            opacity: '0',
            transform: 'translateY(2rem)',
          },
          '100%': {
            opacity: '1',
            transform: 'translateY(0)',
          },
        },
        // ...
      }
      // ...
    }
    // ...
  }

```

Here's what the animation looks like:


https://github.com/reactjs/react.dev/assets/2440089/e3a9b26d-ac41-4575-8cb8-3d7913b30c41

Reply:
@rickhanlonii Done in [`4383ac9` (#6214)](https://github.com/reactjs/react.dev/pull/6214/commits/4383ac921d3eb568b5b7bdc8fc787b82c276a310)

Note that, inside `useEffect` I add an early return when the `errorMessages` doesn't contain `%s`, to avoid extra re-renders for most of the errors.

Reply:
@rickhanlonii Would you like to review the change in [`4383ac9` (#6214)](https://github.com/reactjs/react.dev/pull/6214/commits/4383ac921d3eb568b5b7bdc8fc787b82c276a310)?

In the meantime, I am going to start addressing `<ErrorDecoder>` for MDX you have mentioned.